File: python/camera.py
import math
from ray import Ray
from utils import deg_to_rad
from vector import cross, Point3, Vector

class Camera:
    def __init__(
        self,
        look_from: Point3,
        look_at: Point3,
        v_up: Vector,
        v_field_of_view: float,
        aspect_ratio: float = 16 / 9,
        focal_length: float = 1.0,
    ):
        self.look_from = look_from
        self.look_at = look_at
        self.v_up = v_up
        self.v_field_of_view = float(v_field_of_view)
        self.aspect_ratio = float(aspect_ratio)
        self.focal_length = float(focal_length)

        self.theta: float = deg_to_rad(self.v_field_of_view)
        self.h: float = math.tan(self.theta / 2)
        self.viewport_height: float = 2 * self.h
        self.viewport_width: float = self.aspect_ratio * self.viewport_height
        
        self.w = (self.look_from - self.look_at).unit_vector
        self.u = cross(self.v_up, self.w).unit_vector
        self.v = cross(self.w, self.u)

        self.origin: Point3 = self.look_from
        self.horizontal: Vector = self.u * self.viewport_width
        self.vertical: Vector = self.v * self.viewport_height
        self.lower_left_corner: Vector = self.origin - (self.horizontal / 2) - (self.vertical / 2) - self.w
    
    def get_ray(self, s: float, t: float) -> Ray:
        return Ray(self.origin, self.lower_left_corner + (self.horizontal * s) + (self.vertical * t) - self.origin)

File: rust/src/camera.rs
use crate::vector::Vector;
use crate::point::Point;
use crate::ray::Ray;
use crate::utils::deg_to_rad;


pub struct Camera {
    pub look_from: Point,
    pub look_at: Point,
    pub v_up: Vector,
    
    pub aspect_ratio: f64,
    pub viewport_height: f64,
    pub focal_length: f64,
    pub v_fov: i32,
    horizontal: Vector,
    vertical: Vector,
    lower_left_corner: Point,
}

impl Camera {
    pub fn new(
        look_from: Point,
        look_at: Point,
        v_up: Vector,

        aspect_ratio: f64,
        focal_length: f64,
        v_fov: i32,
    ) -> Camera {
        let theta = deg_to_rad(v_fov);
        let viewport_height = 2.0 * (theta / 2.0).tan();
        let viewport_width = aspect_ratio * viewport_height;

        let w = (look_from - look_at).unit();
        let u = v_up.cross(w).unit();
        let v = w.cross(u);

        let horizontal = viewport_width * u;
        let vertical = viewport_height * v;
        let lower_left_corner = look_from - (horizontal / 2.0) - (vertical / 2.0) - w;

        return Camera {
            look_from,
            look_at,
            v_up,
            aspect_ratio,
            focal_length,
            viewport_height,
            horizontal,
            vertical,
            lower_left_corner,
            v_fov,
        }
    }

    pub fn get_ray(&self, u: f64, v: f64) -> Ray {
        return Ray::new(
            self.look_from,
            self.lower_left_corner + (self.horizontal * u) + (self.vertical * v) - self.look_from,
        );
    }
}